### ISSUE TITLE
1.0.13 not compiling under gcc 4.8.5

### ISSUE THREAD

**Issue by bentilley** [RcppCore/Rcpp#1320]
Hello, thanks for all the great work! I couldn't find anything mentioning this in the existing and closed issues, so hopefully I'm not repeating anything.

#### Issue

As it says on the tin. We're running some projects back in the dark ages on R3 and gcc 4.8.5. The variadic template work added in `1.0.13` means that Rcpp no longer compiles and installs. It's something to do with the voodo of using `sizeof...` in the template argument list of a generic function being used as the default argument to the function parameter of a class method... phew! Here is a minimal reproducible example:

```
// main.cpp
#include <iostream>

template <int n> bool f(int nargs) { return nargs == n; }

class C {
public:
  template <typename... T> void m(bool (*a)(int) = &f<sizeof...(T)>);
};

int main(int argc, char **argv) { return 0; }
```

Compiling this under `g++` `4.8.5` gives me:

```
> g++ -std=gnu++11 main.cpp
main.cpp:7:68: error: expected ‘)’ before ‘(’ token
   template <typename... T> bool m(bool (*valid)(int) = &f<sizeof...(T)>) {
                                                                    ^
main.cpp:7:71: error: expected initializer before ‘>’ token
   template <typename... T> bool m(bool (*valid)(int) = &f<sizeof...(T)>) {
                                                                       ^
```

Here's the output of `g++ -v`:

```
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-44) (GCC)
```

And the output of `sessionInfo()`:

```
R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_3.6.0
```

#### Possible direction of fix?

This is the part of the preprocessor logic that leads to this:

```
// platform/compiler.h:94
	#if GCC_VERSION >= 40800 && __cplusplus >= 201103L
		#define RCPP_USING_CXX11
	#endif
```

The `RCPP_USING_CXX11` macro is what leads us to pick up the new templates, but the new templates don't compile with gcc 4.8 so this is misleading. Probably easiest to fix here to reflect the actual state of older build tool 🤷

Thank you for all the work on the great package 🙏

**Comment by eddelbuettel** [comment]
I will have to close this. We develop, and are measured (by CRAN etc) against *current* versions of tools.  As you say, gcc 4.8.5 is ancient (released in 2015 when gcc 5.2 already existed, but the more relevant date is gcc 4.8.0 released March 2013.  Please just use an Rcpp version from the dark ages.  There are many in the [archives/ folder at CRAN](https://cran.r-project.org/src/contrib/Archive/Rcpp/).

**Comment by eddelbuettel** [comment]
BTW as you are on RedHat, please search previous issues here: your answer is provided by something they offer and call 'devtoolset'.  It will give you more current compilers.
